crates/iota-keys/src/keystore.rs

```diff
@@ -425,62 +533,165 @@ impl FileBasedKeystore {
-        let store = serde_json::to_string_pretty(
-            &self
+    pub fn new(path: &PathBuf) -> Result<Self, anyhow::Error> {
+        if Self::needs_migration(path)? {
+            return Self::migrate_v1_to_v2(path);
+        }
+
+        let (keys, aliases) = if path.exists() {
+            let reader =
+                BufReader::new(File::open(path).with_context(|| {
+                    format!("Cannot open the keystore file: {}", path.display())
+                })?);
+
+            let file: FileBasedKeystoreFile = serde_json::from_reader(reader)
+                .with_context(|| {
+                    format!("Cannot deserialize the keystore file: {}", path.display(),)
+                })
+                .map_err(|e| anyhow!("Invalid keystore file: {}. {}", path.display(), e))?;
+
+            let aliases = file
                 .keys
-                .values()
-                .map(|k| k.encode())
-                .collect::<Result<Vec<_>, _>>()
-                .map_err(|e| anyhow!(e))?,
-        )
-        .with_context(|| format!("Cannot serialize keystore to file: {}", self.path.display()))?;
-        fs::write(&self.path, store)
-            .map_err(|e| anyhow!("Couldn't save keystore to {}: {e}", self.path.display()))?;
-        Ok(())
+                .iter()
+                .map(|aliased| {
+                    let public_key = aliased.key.public();
+                    (
+                        IotaAddress::from(&public_key),
+                        Alias {
+                            alias: aliased.alias.clone(),
+                            public_key_base64: public_key.encode_base64(),
+                        },
+                    )
+                })
+                .collect::<BTreeMap<_, _>>();
+
+            let keys = file
+                .keys
+                .into_iter()
+                .map(|aliased| (IotaAddress::from(&aliased.key.public()), aliased.key))
+                .collect::<BTreeMap<_, _>>();
+
+            (keys, aliases)
+        } else {
+            (BTreeMap::new(), BTreeMap::new())
+        };
+
+        Ok(Self {
+            keys,
+            aliases,
+            path: path.to_path_buf(),
+        })
     }
 
-    pub fn save(&self) -> Result<(), anyhow::Error> {
-        self.save_aliases()?;
-        self.save_keystore()?;
-        Ok(())
+    pub fn set_path(&mut self, path: &Path) {
+        self.path = path.to_path_buf();
     }
 
-    pub fn key_pairs(&self) -> Vec<&IotaKeyPair> {
-        self.keys.values().collect()
+    pub fn save(&self) -> Result<(), anyhow::Error> {
+        let file = FileBasedKeystoreFile {
+            version: 2,
+            keys: self
+                .keys
+                .iter()
+                .map(|(address, key)| AliasedKey {
+                    alias: self
+                        .aliases
+                        .get(address)
+                        .map_or_else(|| self.create_alias(None).unwrap(), |a| a.alias.clone()),
```

Comment:
Is it not possible to not have an alias?

Reply:
No, if you don't provide one, a random one is always generated. The only edge-case where no alias was supported previously was when you had a legacy .keystore file and no .aliases file, so aliases were created for those when you loaded up the keystore.